src/main/java/ch/njol/skript/lang/SkriptParser.java

```diff
@@ -1285,11 +1292,11 @@ private ParseResult parse_i(String pattern, int i, int j) {
 	 * @return The pattern with %codenames% and a boolean array that contains whether the expressions are plural or not
 	 */
 	@Nullable
-	public static NonNullPair<String, boolean[]> validatePattern(final String pattern) {
-		final List<Boolean> ps = new ArrayList<>();
+	public static NonNullPair<String, NonNullPair<ClassInfo<?>, Boolean>[]> validatePattern(final String pattern) {
```

Comment:
Does this really need to be two non null pairs? Can't we just use a multimap here and/or a list instead? Pairs is not as efficient as lists/maps and in this case a Multimap would work now that you're adding another pair.

Reply:
The pairs here don't act like keys -> values, but rather just two objects grouped up. I'm not retrieving the value using the keys, but rather iterating throughout all the pairs and doing stuff with them. I don't need the extra features that maps provide. I'd argue that this is more efficient due to that